src/lib/jobs.rs

```diff
@@ -0,0 +1,346 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `command`      - Command to execute. Examples: "cargo", "ls", "/bin/bash".
+    /// * `args`         - Arguments to accompany the command. Examples: "--version", "-a", "./file.sh".
+    pub async fn start_command(
+        &self,
+        command: String,
+        args: Vec<String>,
+    ) -> Result<(), RLWServerError> {
+        let (tx_output, rx_output): (Sender<u8>, Receiver<u8>) = mpsc::channel();
+        let (tx_pid, rx_pid): (Sender<u32>, Receiver<u32>) = mpsc::channel();
+
+        // Process job
+        let thread = thread::spawn(move || -> Result<ExitStatus, RLWServerError> {
+            execute_command(command, args, Some(&tx_pid), &tx_output)
+        });
+
+        // Set Process ID
+        let mut pid = self.pid.lock().await;
+        *pid = Some(rx_pid.recv()?);
+        mem::drop(pid); // release guard
+
+        // Set status to Running
+        let mut status = self.status.lock().await;
+        *status = status_response::ProcessStatus::Running(true);
+        mem::drop(status); // release guard
+
+        // Populate stdout/stderr output
+        for rec in rx_output {
+            self.output.lock().await.push(rec);
+        }
+
+        let status = thread
+            .join()
+            .map_err(|e| RLWServerError(format!("Error joining on processing thread {:?}", e)))??;
+
+        // Process finished with signal
+        if let Some(s) = status.signal() {
+            let mut status = self.status.lock().await;
+            *status = status_response::ProcessStatus::Signal(s);
+            return Ok(());
+        }
+
+        // Process finished with exit code
+        if let Some(c) = status.code() {
+            let mut status = self.status.lock().await;
+            *status = status_response::ProcessStatus::ExitCode(c);
+            return Ok(());
+        }
+
+        Err(RLWServerError(
+            "Job processing thread closed before it had finished processing".to_string(),
+        ))
+    }
+
+    /// Kill the requested process.
+    ///
+    /// The output of the kill process will be sent to the same output buffer
+    /// as the process you are signaling to kill. The status of the kill signal
+    /// process will not be stored.
+    ///
+    /// # Arguments
+    ///
+    /// * `pid`    - the process id of the process to kill.
+    /// * `forced` - if true, the signal sent will be SIGKILL, and SIGTERM otherwise.
+    pub async fn stop_command(&self, forced: bool) -> Result<(), RLWServerError> {
+        let (tx_output, rx_output): (Sender<u8>, Receiver<u8>) = mpsc::channel();
+
+        // Get the PID of the job to kill
+        let lock_handle = self.pid.lock().await;
+        let pid = lock_handle
+            .ok_or(RLWServerError(
+                "There is no running process associated with this job".to_string(),
+            ))?
+            .to_string();
+        mem::drop(lock_handle);
```

Comment:
I could see relying on `mem::drop(...)` leading to some odd footguns when there's function exit points between the lock and unlock. It's probably nicer in general to let scoping rules handle this by chaining calls (or using nested scopes where that isn't possible):

```suggestion
        let pid = self.pid
            .lock()
            .await
            .ok_or(RLWServerError(
                "There is no running process associated with this job".to_string(),
            ))?
            .to_string();
```

Reply:
From a developer point of view I prefer seeing an explicit drop over the nested scope syntax but I can definitely see your point on it causing some strange behavior.

Have removed all mem::drop(). Thanks for that.